src/v/iceberg/filesystem_catalog.cc

```diff
@@ -0,0 +1,241 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/filesystem_catalog.h"
+
+#include "iceberg/manifest_entry.h"
+#include "iceberg/table_metadata.h"
+#include "iceberg/table_update_applier.h"
+#include "model/timestamp.h"
+
+namespace iceberg {
+
+namespace {
+catalog::errc to_catalog_errc(metadata_io::errc e) {
+    switch (e) {
+    case metadata_io::errc::shutting_down:
+        return catalog::errc::shutting_down;
+    case metadata_io::errc::timedout:
+        return catalog::errc::timedout;
+    case metadata_io::errc::failed:
+        return catalog::errc::io_error;
+    }
+}
+constexpr std::string_view vhint_filename = "version-hint.text";
+ss::sstring tmeta_filename(int version) {
+    return fmt::format("v{}.metadata.json", version);
+}
+} // namespace
+
+ss::future<checked<table_metadata, catalog::errc>>
+filesystem_catalog::create_table(
+  const table_identifier& table_ident,
+  const schema& s,
+  const partition_spec& spec) {
+    // TODO: at some level (maybe the caller), we need to manually re-assign
+    // determininstic, unique field IDs, and re-assign the partition ids. The
+    // Java library and the REST catalog will reassign these on their end, and
+    // we need to follow suit, lest we end up with a mismatch.
+
+    auto new_schema = s.copy();
+    new_schema.schema_id = schema::id_t{0};
+    auto highest_column_id = new_schema.highest_field_id().value_or(
+      nested_field::id_t{0});
+    auto new_spec = spec.copy();
+    new_spec.spec_id = partition_spec::id_t{0};
+    // NOTE: by convention, the first partition field should be id 1000.
+    auto highest_pid = partition_field::id_t{999};
+    for (const auto& p : new_spec.fields) {
+        highest_pid = std::max(highest_pid, p.field_id);
+    }
+    auto new_sort_order = sort_order{
+      sort_order::id_t{0}, chunked_vector<sort_field>{}};
+
+    // Construct a metadata with all initial values.
+    chunked_vector<sort_order> sort_orders;
+    sort_orders.emplace_back(std::move(new_sort_order));
+    chunked_vector<schema> schemas;
+    schemas.emplace_back(std::move(new_schema));
+    chunked_vector<partition_spec> specs;
+    specs.emplace_back(std::move(new_spec));
+    table_metadata tmeta{
+      .format_version = format_version::v2,
+      .table_uuid = uuid_t::create(),
+      .location = table_location(table_ident),
+      .last_sequence_number = sequence_number{0},
+      .last_updated_ms = model::timestamp::now(),
+      .last_column_id = highest_column_id,
+      .schemas = std::move(schemas),
+      .current_schema_id = schema::id_t{0},
+      .partition_specs = std::move(specs),
+      .default_spec_id = partition_spec::id_t{0},
+      .last_partition_id = highest_pid,
+      .sort_orders = std::move(sort_orders),
+      .default_sort_order_id = sort_order::id_t{0},
```

Comment:
will we eventually want to pass them as an argument to the create function ?

Reply:
Probably not for sort order, unless we explicitly want to sort by Kakfa offset or something. FWIW sort order of 0 is unsorted (maybe I'll add a static value for this)

https://github.com/apache/iceberg/blob/f6cdf94094e6c4a4db5e53f3127a04eb3a3a61b1/api/src/main/java/org/apache/iceberg/SortOrder.java#L42-L43